### PR TITLE
[interp] Outline `mono_interp_calli_nat_dynamic_pinvoke` to reduce stack.

### PR DESCRIPTION
**Pull request by jaykrell**
Now that we have that function, also change from `malloc` to `alloca`.
This is part of https://github.com/mono/mono/pull/16171 and ultimately over 100 bytes were saved.

### CODE REVIEW

File: mono/mini/interp/interp.c [hunk 1]
```diff
@@ -2907,13 +2907,47 @@ static int opcode_counts[512];
 #define frame_objref(frame) o
 #endif
 
+// This function is outlined to help save stack in its caller, on the premise
+// that it is relatively rarely called. This also lets it use alloca.
+static
+MONO_NEVER_INLINE
+void
+mono_interp_calli_nat_dynamic_pinvoke (
+	// Parameters are sorted by name.
```
Comment (BrzVlad) [inline mono/mini/interp/interp.c]: :D

Comment (jaykrell) [reply mono/mini/interp/interp.c]: In the NT kernel locals are sorted by name.
It makes it so you don't have to spend time thinking about it, uh, once you pick the names (possibly picking them so they sort how you like).
Here there were too many to come up with a sensible ordering.

File: mono/mini/interp/interp.c [hunk 2]
```diff
@@ -3252,28 +3280,11 @@ interp_exec_method_full (InterpFrame *frame, ThreadContext *context, FrameClause
...
-				/* Pinvoke call is missing the wrapper. See mono_get_native_calli_wrapper */
-				mspecs = g_new0 (MonoMarshalSpec*, csignature->param_count + 1);
 
-				piinfo = piinfo ? piinfo : g_newa (MonoMethodPInvoke, 1);
-				memset (piinfo, 0, sizeof (*piinfo));
-
-				m = mono_marshal_get_native_func_wrapper (m_class_get_image (imethod->method->klass), csignature, piinfo, mspecs, code);
-
-				for (int i = csignature->param_count; i >= 0; i--)
-					if (mspecs [i])
```
Comment (jaykrell) [inline mono/mini/interp/interp.c]: I removed this null check. The other function does it, if the call is ok.

File: mono/mini/interp/interp.c [hunk 3]
```diff
@@ -3252,28 +3280,11 @@ interp_exec_method_full (InterpFrame *frame, ThreadContext *context, FrameClause
...
-				g_free (mspecs);
-
-				child_frame.imethod = mono_interp_get_imethod (imethod->domain, m, error);
-				mono_error_cleanup (error); /* FIXME: don't swallow the error */
-
-				interp_exec_method (&child_frame, context, error);
+			if (imethod->method->dynamic && csignature->pinvoke) {
+				mono_interp_calli_nat_dynamic_pinvoke (&child_frame, code, context, csignature, error, imethod);
 			} else {
+				const gboolean save_last_error = *(guint16 *)(ip - 3 + 2);
```
Comment (jaykrell) [inline mono/mini/interp/interp.c]: Notice the ` -3` here.  A little hard on the maintainer.

File: mono/mini/interp/interp.c [hunk 4]
```diff
@@ -3230,19 +3260,15 @@ interp_exec_method_full (InterpFrame *frame, ThreadContext *context, FrameClause
...
 			frame->ip = ip;
-			
-			csignature = (MonoMethodSignature*)imethod->data_items [* (guint16 *)(ip + 1)];
-			save_last_error = *(guint16 *)(ip + 2);
+
+			MonoMethodSignature* csignature = (MonoMethodSignature*)imethod->data_items [* (guint16 *)(ip + 1)];
+
 			ip += 3;
 			--sp;
-			--endsp;
```
Comment (jaykrell) [inline mono/mini/interp/interp.c]: Bad merge here with another change lost this line.

File: mono/mini/interp/interp.c (hunk 2 above)
Comment (jaykrell) [reply mono/mini/interp/interp.c]: Restored.

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Pull request merged**